receiver/receivercreator/hints.go

```diff
@@ -0,0 +1,223 @@
+	logger *zap.Logger
+	config K8sHintsConfig
+}
+
+// createScraperTemplateFromHints creates a receiver configuration based on the provided hints.
+// Hints are extracted from Pod's annotations.
+// Scraper configurations are only created for Port Endpoints.
+// TODO: container_logs configurations are only created for Pod Container Endpoints.
+func (builder *K8sHintsBuilder) createScraperTemplateFromHints(env observer.EndpointEnv) (*receiverTemplate, error) {
+	var endpointType string
+	var podUID string
+	var annotations map[string]string
+
+	endpointType = getStringEnv(env, "type")
+	if endpointType == "" {
+		return nil, fmt.Errorf("could not get endpoint type: %v", zap.Any("env", env))
+	}
+
+	if endpointType != string(observer.PortType) {
+		return nil, nil
+	}
+
+	builder.logger.Debug("handling hints for added endpoint", zap.Any("env", env))
+
+	if pod, ok := env["pod"]; ok {
+		endpointPod, ok := pod.(observer.EndpointEnv)
+		if !ok {
+			return nil, fmt.Errorf("could not extract endpoint's pod: %v", zap.Any("endpointPod", pod))
+		}
+		ann := endpointPod["annotations"]
+		if ann != nil {
+			annotations, ok = ann.(map[string]string)
+			if !ok {
+				return nil, fmt.Errorf("could not extract annotations: %v", zap.Any("annotations", ann))
+			}
+			if len(annotations) == 0 {
+				return nil, nil
+			}
+		}
+		podUID = endpointPod["uid"].(string)
+	} else {
+		return nil, nil
+	}
+
+	if builder.config.Enabled {
+		// Only handle Endpoints of type port for scraper
+		return builder.createScraper(annotations, env, podUID)
+	}
+	return nil, nil
+}
+
+func (builder *K8sHintsBuilder) createScraper(
+	annotations map[string]string,
+	env observer.EndpointEnv,
+	podUID string) (*receiverTemplate, error) {
+
+	var port uint16
+	portName := getStringEnv(env, "name")
+
+	if !discoveryEnabled(annotations, portName) {
+		return nil, nil
+	}
+
+	subreceiverKey := getHintAnnotation(annotations, scraperHint, portName)
+	if subreceiverKey == "" {
+		// no scraper hint detected
+		return nil, nil
+	}
+	builder.logger.Debug("handling added hinted receiver", zap.Any("subreceiverKey", subreceiverKey))
+
+	defaultEndpoint := getStringEnv(env, "endpoint")
+	userConfMap := getConfFromAnnotations(annotations, defaultEndpoint, portName)
+
+	signalsAnn := getHintAnnotation(annotations, signalsHint, portName)
+	signals := getSignalsConf(signalsAnn)
+	userConfMap["signals"] = signals
+
+	if p, ok := env["port"]; ok {
+		port = p.(uint16)
+		if port == 0 {
+			return nil, fmt.Errorf("could not extract port: %v", zap.Any("env", env))
+		}
+	} else {
+		return nil, fmt.Errorf("could not extract port: %v", zap.Any("env", env))
+	}
+	subreceiver, err := newReceiverTemplate(fmt.Sprintf("%v/%v_%v", subreceiverKey, podUID, port), userConfMap)
+	if err != nil {
+		builder.logger.Error("error adding subreceiver", zap.Any("err", err))
+		return nil, err
+	}
+
+	builder.logger.Debug("adding hinted receiver", zap.Any("subreceiver", subreceiver))
+	return &subreceiver, nil
```

Comment:
The following log messages are already covered by the caller function, maybe we can simplify it with:

```suggestion
	return newReceiverTemplate(fmt.Sprintf("%v/%v_%v", subreceiverKey, podUID, port), userConfMap)
```
(newReceiverTemplate needs to be changed to return a reference)